### PR TITLE
Feature 2416 pulse high scores

### PR DESCRIPTION
**Pull request by mvolkmann**

### CODE REVIEW

File: server/src/main/resources/db/dev/R__Load_testing_data.sql [hunk 1]
```diff
@@ -34,6 +34,16 @@ INSERT INTO member_profile -- Gina Bremehr
 VALUES
 ('01b7d769-9fa2-43ff-95c7-f3b950a27bf9', PGP_SYM_ENCRYPT('Gina','${aeskey}'), PGP_SYM_ENCRYPT('Bremehr','${aeskey}'), PGP_SYM_ENCRYPT('COO','${aeskey}'), PGP_SYM_ENCRYPT('St. Louis','${aeskey}'), PGP_SYM_ENCRYPT('bremehrg@objectcomputing.com','${aeskey}'), '12312345', '2012-09-20', PGP_SYM_ENCRYPT('Epitome of Strong Woman','${aeskey}'), null, '1988-09-21');
 
+INSERT INTO member_profile -- Big Boss
```
Comment (mvolkmann) [inline server/src/main/resources/db/dev/R__Load_testing_data.sql:37]: I changed some `member_profile` records so they have `supervisorid` values. This was needed to test some if the changes for this story.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 2]
```diff
@@ -16,13 +16,15 @@ import {
 import { Comment } from '@mui/icons-material';
 import {
   Avatar,
-  Button,
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:19]: Not used.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 3]
```diff
@@ -31,7 +33,6 @@ import { DatePicker } from '@mui/x-date-pickers/DatePicker';
 import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
 
 import { getAvatarURL, resolve } from '../api/api.js';
-import Pulse from '../components/pulse/Pulse.jsx';
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:34]: Not used.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 4]
```diff
@@ -51,6 +52,18 @@ const ociDarkBlue = '#2c519e';
 // const ociOrange = '#f8b576'; // too light
 const orange = '#b26801';
 
+const ScoreOption = {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:55]: When viewing the 5 highest and lowest average scores, the user can choose between viewing only internal scores, only external scores, or a combination of internal and external scores.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 5]
```diff
@@ -121,34 +136,74 @@ const PulseReportPage = () => {
   }, [dateFrom, dateTo]);
   */
 
+  const average = (arr) => arr.reduce((a, b) => a + b, 0) / arr.length;
+
   // This creates data in the format that recharts needs from pulse data.
   useEffect(() => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:142]: Lots of data munging in this `useEffect` to prepare data for the line chart, the bar chart, and the five highest and lowest average scores.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 6]
```diff
@@ -194,73 +248,67 @@ const PulseReportPage = () => {
     //      because this permission has not been implemented yet.
     // if (selectHasPulseReportPermission(state)) loadPulses();
     loadPulses();
-    setTeamMembers(state.memberProfiles);
   }, [csrf, dateFrom, dateTo]);
 
-  const handleDateFromChange = dayJsDate => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:200]: These functions were not deleted, they were just moved down.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 7]
```diff
@@ -194,73 +248,67 @@ const PulseReportPage = () => {
...
-    const teamMemberIds = teamMembers.map(member => member.id);
-    if (teamMemberIds.length) {
-      filteredPulses = pulses.filter(pulse =>
-        teamMemberIds.includes(pulse.teamMemberId))
-    }
+  useEffect(() => {
+    setTeamMembers(state.memberProfiles || []);
+  }, [csrf, state]);
 
+  const averageRow = scores => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:257]: This returns a single table row to be rendered in the tables of highest and lowest average scores.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 8]
```diff
@@ -307,6 +355,50 @@ const PulseReportPage = () => {
     </Card>
   );
 
+  const handleCommentClick = (pulse) => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:358]: When the comment icon at the end of a row of pulse data is clicked, this function is called.
It triggers displaying the internal and external comments associated with that pulse entry in a modal dialog.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 9]
```diff
@@ -307,6 +355,50 @@ const PulseReportPage = () => {
...
+    if (date > dateTo) setDateTo(date);
+  };
+
+  const handleDateToChange = dayJsDate => {
+    const date = new Date(dayJsDate.valueOf());
+    if (date < dateFrom) setDateFrom(date);
+    setDateTo(date);
+  };
+
+  const handleTeamMembersChange = members => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:375]: This is used to change the set of team members selected in the `MemberSelector` component.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 10]
```diff
@@ -307,6 +355,50 @@ const PulseReportPage = () => {
...
+    setTeamMembers(members);
+  };
+
+  const scoreCard = (highest) => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:379]: This renders a table of either the highest five average scores or the lowest five average scores based on the boolean passed in.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 11]
```diff
@@ -345,6 +437,52 @@ const PulseReportPage = () => {
     </Card>
   );
 
+  const responseSummary = () => {
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:440]: This renders the responses that were used to drive the bar chart.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 12]
```diff
@@ -376,25 +514,25 @@ const PulseReportPage = () => {
             selected={teamMembers}
           />
           {lineChart()}
+          {averageScores()}
           {barChart()}
+          <Modal open={showComments} onClose={() => setShowComments(false)}>
```
Comment (mvolkmann) [inline web-ui/src/pages/PulseReportPage.jsx:519]: This dialog displays the internal and external comments associated with a selected pulse entry (by click its comment icon).

**Review by S78901 — APPROVED** [review]
Looks great.

File: web-ui/src/pages/PulseReportPage.jsx [hunk 13]
```diff
@@ -121,34 +136,74 @@ const PulseReportPage = () => {
...
-    setLineChartData(data);
+    setLineChartData(lineChartData);
     setBarChartData(frequencies);
-  }, [pulses, teamMembers]);
+
+    for (const memberId of Object.keys(averageData)) {
+      const averages = averageData[memberId];
+      averages.externalAverage = average(averages.externalScores);
+      averages.internalAverage = average(averages.internalScores);
+      averages.combinedAverage = average([...averages.externalScores, ...averages.internalScores]);
```
Comment (vhscom) [inline web-ui/src/pages/PulseReportPage.jsx:203]: Very elegant.

**Review by vhscom — APPROVED** [review]
LGTM. Very nice!

**Pull request merged**